core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -1078,8 +1078,12 @@ private[spark] object JsonProtocol {
         val blockManagerAddress = blockManagerIdFromJson(json \ "Block Manager Address")
         val shuffleId = (json \ "Shuffle ID").extract[Int]
         val mapId = (json \ "Map ID").extract[Long]
-        val mapIndex = (json \ "Map Index") match {
-          case JNothing => 0
+        val mapIndex = json \ "Map Index" match {
+          case JNothing =>
+            // Note, we use the invalid value -2 here to fill the map index for backward
+            // compatibility. Otherwise, the fetch failed event will be dropped when the history
+            // server loads the event log written by the Spark version before 3.0.
+            -2
```

Comment:
Any special reasons using `-2` instead of `-1`? BTW, I'm unsure whether this filed will be displayed in UI, if so, we better show `Unknown` for invalid value in UI.

Reply:
Yeah once we provide it with magic number to mark INVALID/UNKNOWN, there should be relevant code change to handle INVALID/UNKNOWN value. Looks like the code change doesn't address it, which makes me unsure how it will be represented on the UI.

Reply:
@HeartSaVioR I think it is represented as error string in `FetchFailed.toErrorString`. So we just need to make sure the default unknown value is consistent in the code here and `FetchFailed.toErrorString`. Personally I prefer `Int.MinValue` since `-1` can't be used here.

Reply:
Thanks Gengliang for confirming!
```
Any special reasons using -2 instead of -1?
```
mapIndex = -1 has been used in `MetadataFetchFailedException`.

Yep, I think `Int.MinValue` is a better choice.

Reply:
Thanks for the info! If the value is only leveraged via `FetchFailed.toErrorString`, what we may want to do (and proposed in previous comment) is projecting the magic number to the specific string so that it can be determined easily. +1 to `Int.MinValue`.

Reply:
And I'm seeing it's done. Thanks for the quick address!

Reply:
Yep, already addressed in 9dfa5c0